Break Differential up to make it testable

sectioneight opened the pull request:
Previously was mixing concerns with both talking to arcanist and
presenting results. By breaking out a DifferentialClient that I can
easily mock, we can test the core functionality without a ton of boilerplate.

sectioneight commented:
@jjx when you have a spare moment

sectioneight commented on src/test/java/com/uber/jenkins/phabricator/conduit/DifferentialTest.java:
```diff
@@ -0,0 +1,80 @@
+    Differential d;
+
+    public DifferentialTest() throws InterruptedException, ArcanistUsageException, IOException {
+        d = diffWithResponse();
```
not totally clear on how `@Rule` works, but it wouldn't let me use this up at the definition because it throws exceptions. does this get setup / torn down for each test run, or am i leaking state? (doesn't matter in this case since the object is immutable, but for my own edification)

jjx replied:
Use @ Before to set up instead.

jjx replied:
http://junit.sourceforge.net/javadoc/org/junit/Before.html

sectioneight commented:
@jjx bump

jjx commented on src/main/java/com/uber/jenkins/phabricator/conduit/DifferentialClient.java:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    protected JSONObject callConduit(String methodName, Map<String, String> params) throws IOException, InterruptedException, ArcanistUsageException {
```
Move this method to after all the public's

jjx commented on src/main/java/com/uber/jenkins/phabricator/conduit/DifferentialClient.java:
```diff
@@ -0,0 +1,122 @@
+        params.put("type", pass ? "pass" : "fail");
+        params.put("buildTargetPHID", phid);
+
+        this.callConduit("harbormaster.sendmessage", params);
```
Is this result any useful? Want to return it?

jjx commented on src/test/java/com/uber/jenkins/phabricator/conduit/DifferentialTest.java:
```diff
@@ -0,0 +1,80 @@
+
+    private Differential diffWithResponse() throws InterruptedException, ArcanistUsageException, IOException {
+        DifferentialClient client = mock(DifferentialClient.class);
+        when(client.fetchDiff()).thenReturn(getValidQueryResponse());
```
Can you mock at the callConduit layer instead? You can test more logic from fetchDiff that way. This might be useful:

@Test(expected = Exception.class)
public void testCoolTestThatExpectsException() { ... }

jjx commented on src/test/java/com/uber/jenkins/phabricator/conduit/DifferentialTest.java:
```diff
@@ -0,0 +1,80 @@
+// THE SOFTWARE.
+
+package com.uber.jenkins.phabricator.conduit;
+
```
Extra newline

sectioneight replied:
Fixed

jjx commented on src/main/java/com/uber/jenkins/phabricator/conduit/DifferentialClient.java:
```diff
@@ -0,0 +1,122 @@
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public void harbormaster(String phid, boolean pass) throws IOException, InterruptedException, ArcanistUsageException {
```
Can we rename this method to something more meaningful?

jjx commented on src/main/java/com/uber/jenkins/phabricator/conduit/DifferentialClient.java:
```diff
@@ -0,0 +1,122 @@
+        try {
+            response = query.getJSONObject("response");
+        } catch (JSONException e) {
+            e.printStackTrace();
```
Hmm... won't this just go to the jenkins process' stdout/stderr? That won't be that useful

sectioneight replied:
It shows up in the jenkins logs, but yea, I will remove it

jjx commented on src/main/java/com/uber/jenkins/phabricator/conduit/DifferentialClient.java:
```diff
@@ -0,0 +1,122 @@
+    private final LauncherFactory launcher;
+    private final String diffID;
+
+    public DifferentialClient(String diffID, LauncherFactory launcher, String conduitToken, String arcPath) {
```
Can these be ordered?

sectioneight replied:
What do you mean by "ordered"?

jjx replied:
Constructor and declaration order.

sectioneight replied:
ah, yes, of course, i definitely know how to program :)

jjx commented on src/main/java/com/uber/jenkins/phabricator/conduit/Differential.java:
```diff
@@ -33,44 +32,15 @@
-                            response.toString(2)));
-
-        }
+    public Differential(DifferentialClient client) throws ArcanistUsageException, IOException, InterruptedException {
```
Let's just take in the differential json here. Taking in the client and caching it is not necessary. Let's just make differential a model object that can fetch various pieces of data.

sectioneight replied:
Nice. Done.

jjx commented on src/test/java/com/uber/jenkins/phabricator/conduit/DifferentialTest.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private Differential diffWithResponse() throws InterruptedException, ArcanistUsageException, IOException {
+        DifferentialClient client = mock(DifferentialClient.class);
```
At some point, we're probably going to add a MockDifferentialClient class to the test package and have some sort of response queue object in there that returns expected json blobs when we do callConduit. But separating the differential client and differential model is a HUGE step! This is some sweet stuff!! :)

jjx commented on src/test/java/com/uber/jenkins/phabricator/conduit/DifferentialTest.java:
```diff
@@ -0,0 +1,80 @@
+import java.io.InputStream;
+
+public class DifferentialTest extends TestCase {
+    private static final String fakeDiffID = "not-a-real-id";
```
In android, we usually do caps with underscores for static finals:
fakeDiffID -> FAKE_DIFF_ID

Also:
Differential d -> Differential differential

sectioneight commented:
@jjx addressed all your comments so I'm going to merge, let me know if there's more stuff you'd like me to follow up on and I'm happy to address, just trying to get master integrated so we don't conflict.
